File: Algorithm/DataStructure/FenwickTree/fenwick_tree.rs
use std::{
    ops::{Add, AddAssign},
    process::Output,
};

pub struct FenwickTree<T: Copy> {
    data: Vec<T>,
}

impl<T: Copy> FenwickTree<T> {
    pub fn new(n: usize, init_val: T) -> Self {
        Self {
            data: vec![init_val; n],
        }
    }

    pub fn add(&mut self, mut p: usize, val: T)
    where
        T: Add<Output = T> + AddAssign,
    {
        p += 1;
        while p <= self.data.len() {
            self.data[p - 1] += val;
            p += p & (!p + 1);
        }
    }

    pub fn sum(&self, mut p: usize, mut init_val: T) -> T
    where
        T: Add<Output = T> + AddAssign,
    {
        p += 1;
        while p > 0 {
            init_val += self.data[p - 1];
            p -= p & (!p + 1);
        }

        init_val
    }
}

File: Prev-Repo/Algorithm/DataStructure/FenwickTree/FenwickTree.cs
public class FenwickTree<T>
    where T : System.Numerics.IAdditionOperators<T, T, T>,
              System.Numerics.ISubtractionOperators<T, T, T>,
              new()
{
    private T[] _data;

    public int Length => _data.Length;

    public FenwickTree(int length)
    {
        _data = new T[length];
        for (int i = 0; i < length; i ++)
            _data[i] = new();
    }

    public void Add(int index, in T value)
    {
        index ++;
        for (int i = index; i < Length; i += i & -i)
        {
            _data[i - 1] += value;
        }
    }

    public T Sum(int index)
    {
        index ++;
        T result = new();
        for (int i = index; i > 0; i -= i & -i)
        {
            result += _data[i - 1];
        }
        return result;
    }

    public T RangeSum(int l, int r) => Sum(r) - Sum(l - 1);
}